[Enhancement] Add universal replace_or_append function

redhatrises opened the pull request:
- Adds universal replace_or_append function to remediation_functions
- Provide a couple of examples of usage in existing remediation scripts
- Fixes #611
- Fixes #548

redhatrises commented:
Please note that the final iteration of this is some sort of XSLT script that will replace the CCENUM variable with the actual CCE value listed in the XCCDF. As I am really bad at XSLT, I am not sure how to accomplish grabbing the CCE value from the XCCDF and replacing CCENUM with the CCE value. If anyone can whip something up quickly to help with that, that would be great; otherwise, I will open an issue for it.

iankko commented on shared/remediations/bash/templates/remediation_functions:
```diff
@@ -701,3 +701,86 @@ do
 
 }
+
+# Function to enable/disable and start/stop services on RHEL and Fedora systems.
```
This should be updated to reflect purpose of 'replace_or_append' function.

redhatrises commented:
Oopps. My bad. @iankko fixed.

iankko commented:
I will have a look at this one once Jenkins finished.

iankko commented:
I have tested this on both RHEL-6 && RHEL-7 => couple of observations below (or in the particular code line changes in the upcoming comments).

* First the changeset is adding the second version of ```selinux_policytype.sh``` script:
```
$ find shared/ RHEL/7 -name selinux_policytype.sh
shared/remediations/bash/selinux_policytype.sh
RHEL/7/input/remediations/bash/selinux_policytype.sh
```
and 
```
$ find shared/ RHEL/7 -name selinux_state.sh
shared/remediations/bash/selinux_state.sh
RHEL/7/input/remediations/bash/selinux_state.sh
```
Was that intentional? Because besides other what this means that ```shared/``` (e.g. yours) implementation of those two remediation scripts won't be used, till the ```RHEL/7``` specific version is deleted first.

Tested also on ```RHEL/6``` system (where there aren'ts RHEL/6 specific remediations for these two checks) to see if the implementation works, and found the following (more in the direct code lines in the comments below).

iankko commented on shared/remediations/bash/templates/remediation_functions:
```diff
@@ -701,3 +701,87 @@ do
+  # Test if the config_file is a symbolic link. If so, use --follow-symlinks with sed.
+  # Otherwise, regular sed command will do.
+  if test -L $config_file; then
+    sed_command="sed --follow-symlinks"
```
Second: Should the ```sed_command``` here contain ```-i``` argument (otherwise sed won't change the file in question in place).

iankko commented on shared/remediations/bash/templates/remediation_functions:
```diff
@@ -701,3 +701,87 @@ do
+  if test -L $config_file; then
+    sed_command="sed --follow-symlinks"
+  else
+    sed_command="sed"
```
Second: Ditto here -- IMHO you wanted ```-i``` sed option here.

iankko commented on shared/remediations/bash/templates/remediation_functions:
```diff
@@ -701,3 +701,87 @@ do
+
+  # If the key exists, change it. Otherwise, add it to the config_file.
+  if grep -qi $key $config_file ; then
+    $sed_command "s/$key.*/$formatted_output/g" $config_file
```
Since the ```-i``` option is missing here, even in the case the desired ```$key``` is found in the ```$config_file``` the change won't be applied (the corrected file output -- how it would look like after applying the change would be printed out to stdout). IMHO you want the ```-i``` option either in this row, or in both of the ```$sed_command``` definitions above.

iankko commented on shared/remediations/bash/templates/remediation_functions:
```diff
@@ -701,3 +701,87 @@ do
+  if grep -qi $key $config_file ; then
+    $sed_command "s/$key.*/$formatted_output/g" $config_file
+  else
+#    echo -ne "\n# Per $cce: Set $formatted_output in $config_file" >> $config_file
```
Last (but not least): This one looks like temporary commented out code, which has been forgotten to be un-commented. And the two code statements at lines #782 and #783 were expected to be deleted IMHO in the final version. Because without doing these changes you again end up just printing the changes (the modified content of the final file) to the stdout, rather than to actually inserting the changes into the file.

iankko commented:
Once the above changes are done the implementation should work (but didn't explicitly test that).

redhatrises commented:
@iankko I did not look to see if the selinux remediation scripts exists under `RHEL/7` as remediation scripts resided under `shared/remediation/bash` as well as the OVAL checks under `shared/oval`. Fixed this. 

Also apparently (as you found) did not remove the testing comments/sed variables which was stupid of me. This is also fixed.

iankko commented:
The first sub-case
* (when particular $key to grep $config_file for is already present in the $config_file) works fine for the case the remediation is called directly from the command line:
```
# cat /etc/redhat-release 
Red Hat Enterprise Linux Server release 7.2 (Maipo)

# cat /etc/selinux/config 

# This file controls the state of SELinux on the system.
# SELINUX=
#     enforcing - SELinux security policy is enforced.
#     permissive - SELinux prints warnings instead of enforcing.
#     disabled - No SELinux policy is loaded.
SELINUX=disabled
# SELINUXTYPE=
#     targeted - Targeted processes are protected,
#     minimum - Modification of targeted policy. Only selected processes are protected. 
#     mls - Multi Level Security protection.
SELINUXTYPE=mls

# bash --norc selinux_policytype.sh 
var_selinux_policy_name is not defined. Exiting.
# export var_selinux_policy_name='targeted'
# bash --norc selinux_policytype.sh
# cat /etc/selinux/config 

# This file controls the state of SELinux on the system.
# SELINUX=
#     enforcing - SELinux security policy is enforced.
#     permissive - SELinux prints warnings instead of enforcing.
#     disabled - No SELinux policy is loaded.
SELINUX=disabled
# SELINUXTYPE=
#     targeted - Targeted processes are protected,
#     minimum - Modification of targeted policy. Only selected processes are protected. 
#     mls - Multi Level Security protection.
SELINUXTYPE=targeted
```
But <b>does not work properly</b> when the remediation is attempted directly from the benchmark:
```
# cat /etc/selinux/config 

# This file controls the state of SELinux on the system.
# SELINUX=
#     enforcing - SELinux security policy is enforced.
#     permissive - SELinux prints warnings instead of enforcing.
#     disabled - No SELinux policy is loaded.
SELINUX=disabled
# SELINUXTYPE=
#     targeted - Targeted processes are protected,
#     minimum - Modification of targeted policy. Only selected processes are protected. 
#     mls - Multi Level Security protection.
SELINUXTYPE=mls
```
Then perform the remediation either by calling oscap --remediate or scap-workbench (having 'Remediate' check-box checked in). This results into the following situation:
```
# cat /etc/selinux/config 

# This file controls the state of SELinux on the system.
# SELINUX=
#     enforcing - SELinux security policy is enforced.
#     permissive - SELinux prints warnings instead of enforcing.
#     disabled - No SELinux policy is loaded.
SELINUX=disabled
# SELINUXTYPE=
#     targeted - Targeted processes are protected,
#     minimum - Modification of targeted policy. Only selected processes are protected. 
#     mls - Multi Level Security protection.
targeted=
```
Same result is visible on RHEL-6 when running the remediation via benchmark directly.
* Didn't test the second sub-case because it does not modify the real config, just prints out to stdout what would be written into that config (IOW incorrect two lines have been removed in https://github.com/redhatrises/scap-security-guide/commit/1caf651a1f880c48d1d032d957349300563184e0 change).

redhatrises commented:
@iankko sorry for the noise on this review. I have tested it. Issue appears to have been with an inline sed.

iankko commented:
Verified this version works fine on both RHEL-6 && RHEL-7 systems in both of the cases below:
* $key value is present in the $config_file already, but set to different value than expected by the requirement,
* $key value is not present in the $config_file yet.

ACK from me for merging of this one (based on the testing && above comment).

iankko commented:
@redhatrises 
> @iankko sorry for the noise on this review. I have tested it. Issue appears to have been with an inline sed.

No worries! The main thing is we have caught the basic functionality problems of this new feature yet before it's started to be used across the content more widely.

Cases like this are examples why it's necessary to provide also examples how the new feature should be used / utilized while introducing it (so people can test the proposal).

Anyway thank you for the time spent && work on this! (and that all of those my review comments didn't scare you :)). This is very handy function to be available (I will let the future to prove it :)).
